File: portfolio_optimizer.cpp
#include <iostream>
#include <string>
#include <vector>
#include <map>
#include <set>
#include <random>
#include <algorithm>
#include <cmath>
#include <sstream>
#include <iomanip>
#include "prism_client.h"

// Simple JSON-like structure using standard C++ types
struct PortfolioContext {
    std::string message;
    double budget;
    int age;
    double salary;
    std::string start_date;
    std::string end_date;
    std::set<std::string> disliked_sectors;
};

class PortfolioOptimizer {
private:
    PortfolioContext context;
    std::map<std::string, double> risk_factors = {
        {"age", 0.5},        // Higher age = more conservative
        {"employment_status", 0.3},  // Unemployed = more conservative
        {"income", 0.2}      // Lower income = more conservative
    };
    std::map<std::string, double> sector_weights;
    double budget = 0;
    std::set<std::string> disliked_stocks;
    double risk_tolerance = 0.5;  // Default risk tolerance
    std::set<std::string> disliked_sectors;
    std::string start_date;
    std::string end_date;
    double salary = 0;
    int age = 0;

public:
    PortfolioOptimizer() = default;

    void parseContext(const std::string& contextStr) {
        context.message = contextStr;
        
        // Extract client preferences from message
        std::string message = contextStr;
        
        // Parse age
        size_t agePos = message.find("is ") + 3;
        size_t yearsPos = message.find(" years", agePos);
        if (agePos != std::string::npos && yearsPos != std::string::npos) {
            try {
                std::string ageStr = message.substr(agePos, yearsPos - agePos);
                std::cout << "Parsing age from: '" << ageStr << "'" << std::endl;
                age = std::stoi(ageStr);
            } catch (const std::exception& e) {
                std::cerr << "Error parsing age: " << e.what() << std::endl;
            }
        }
        
        // Parse budget
        size_t budgetPos = message.find("budget of $") + 10;
        size_t dotPos = message.find(".", budgetPos);
        if (budgetPos != std::string::npos && dotPos != std::string::npos) {
            try {
                std::string budgetStr = message.substr(budgetPos, dotPos - budgetPos);
                // Remove any non-numeric characters except decimal point
                budgetStr.erase(std::remove_if(budgetStr.begin(), budgetStr.end(), 
                    [](char c) { return !std::isdigit(c) && c != '.'; }), budgetStr.end());
                std::cout << "Parsing budget from: '" << budgetStr << "'" << std::endl;
                budget = std::stod(budgetStr);
            } catch (const std::exception& e) {
                std::cerr << "Error parsing budget: " << e.what() << std::endl;
            }
        }
        
        // Parse dates
        size_t startDatePos = message.find("Started investing on ") + 20;
        size_t andPos = message.find(" and", startDatePos);
        if (startDatePos != std::string::npos && andPos != std::string::npos) {
            start_date = message.substr(startDatePos, andPos - startDatePos);
            // Clean up any extra spaces
            start_date.erase(0, start_date.find_first_not_of(" \t"));
            start_date.erase(start_date.find_last_not_of(" \t") + 1);
        }
        
        size_t endDatePos = message.find("ended on ") + 9;
        size_t endDotPos = message.find(".", endDatePos);
        if (endDatePos != std::string::npos && endDotPos != std::string::npos) {
            end_date = message.substr(endDatePos, endDotPos - endDatePos);
            // Clean up any extra spaces
            end_date.erase(0, end_date.find_first_not_of(" \t"));
            end_date.erase(end_date.find_last_not_of(" \t") + 1);
        }
        
        // Parse salary
        size_t salaryPos = message.find("salary of $") + 10;
        size_t salaryDotPos = message.find(".", salaryPos);
        if (salaryPos != std::string::npos && salaryDotPos != std::string::npos) {
            try {
                std::string salaryStr = message.substr(salaryPos, salaryDotPos - salaryPos);
                // Remove any non-numeric characters except decimal point
                salaryStr.erase(std::remove_if(salaryStr.begin(), salaryStr.end(), 
                    [](char c) { return !std::isdigit(c) && c != '.'; }), salaryStr.end());
                std::cout << "Parsing salary from: '" << salaryStr << "'" << std::endl;
                salary = std::stod(salaryStr);
            } catch (const std::exception& e) {
                std::cerr << "Error parsing salary: " << e.what() << std::endl;
            }
        }
        
        // Parse disliked sectors
        if (message.find("avoids") != std::string::npos) {
            size_t avoidsPos = message.find("avoids ") + 7;
            size_t avoidsDotPos = message.find(".", avoidsPos);
            if (avoidsPos != std::string::npos && avoidsDotPos != std::string::npos) {
                std::string sectorsStr = message.substr(avoidsPos, avoidsDotPos - avoidsPos);
                // Split sectors by comma and add to set
                size_t pos = 0;
                while ((pos = sectorsStr.find(',')) != std::string::npos) {
                    std::string sector = sectorsStr.substr(0, pos);
                    sector.erase(0, sector.find_first_not_of(" \t"));
                    sector.erase(sector.find_last_not_of(" \t") + 1);
                    disliked_sectors.insert(sector);
                    sectorsStr.erase(0, pos + 1);
                }
                if (!sectorsStr.empty()) {
                    sectorsStr.erase(0, sectorsStr.find_first_not_of(" \t"));
                    sectorsStr.erase(sectorsStr.find_last_not_of(" \t") + 1);
                    disliked_sectors.insert(sectorsStr);
                }
            }
        }
        
        // Calculate risk tolerance based on client profile
        double age_factor = 1.0 - (std::min(1.0, static_cast<double>(age) / 100.0));  // Invert age factor
        double employment_factor = 0.7;  // Higher for employed
        double income_factor = std::min(1.0, salary / 100000.0);  // Normalize salary to 0-1 range
        
        risk_tolerance = (
            age_factor * risk_factors["age"] +
            employment_factor * risk_factors["employment_status"] +
            income_factor * risk_factors["income"]
        );
        
        // Ensure risk tolerance is between 0.3 and 0.8
        risk_tolerance = std::max(0.3, std::min(0.8, risk_tolerance));
        
        std::cout << "\nParsed context:" << std::endl;
        std::cout << "Age: " << age << std::endl;
        std::cout << "Budget: $" << budget << std::endl;
        std::cout << "Salary: $" << salary << std::endl;
        std::cout << "Start Date: " << start_date << std::endl;
        std::cout << "End Date: " << end_date << std::endl;
        std::cout << "Disliked Sectors: ";
        for (const auto& sector : disliked_sectors) {
            std::cout << sector << " ";
        }
        std::cout << std::endl;
        std::cout << "Risk Tolerance: " << risk_tolerance << std::endl;
    }

    std::vector<std::pair<std::string, double>> calculateStockWeights(const std::string& strategy = "balanced") {
        if (context.message.empty()) {
            throw std::runtime_error("Context not initialized");
        }
        
        // Define sector classifications
        std::vector<std::string> conservative_sectors = {"Utilities", "Consumer Staples", "Healthcare"};
        std::vector<std::string> aggressive_sectors = {"Technology", "Consumer Discretionary", "Communication Services"};
        std::vector<std::string> moderate_sectors = {"Financials", "Industrials", "Materials"};
        
        // Sample stocks for each category
        std::vector<std::pair<std::string, double>> conservative_stocks = {
            {"PG", 1}, {"JNJ", 1}, {"KO", 1}
        };
        std::vector<std::pair<std::string, double>> moderate_stocks = {
            {"JPM", 1}, {"MMM", 1}, {"CAT", 1}
        };
        std::vector<std::pair<std::string, double>> aggressive_stocks = {
            {"AAPL", 1}, {"MSFT", 1}, {"GOOGL", 1}
        };
        
        std::vector<std::pair<std::string, double>> portfolio;
        double conservative_weight = 0.0;
        double moderate_weight = 0.0;
        double aggressive_weight = 0.0;
        
        if (strategy == "balanced") {
            if (risk_tolerance < 0.5) {
                conservative_weight = 0.6;
                moderate_weight = 0.3;
                aggressive_weight = 0.1;
            } else if (risk_tolerance > 0.7) {
                conservative_weight = 0.2;
                moderate_weight = 0.3;
                aggressive_weight = 0.5;
            } else {
                conservative_weight = 0.4;
                moderate_weight = 0.4;
                aggressive_weight = 0.2;
            }
        } else if (strategy == "aggressive") {
            conservative_weight = 0.1;
            moderate_weight = 0.2;
            aggressive_weight = 0.7;
        } else if (strategy == "conservative") {
            conservative_weight = 0.8;
            moderate_weight = 0.15;
            aggressive_weight = 0.05;
        } else if (strategy == "tech_heavy") {
            conservative_weight = 0.1;
            moderate_weight = 0.1;
            aggressive_weight = 0.8;
        } else if (strategy == "equal_weight") {
            size_t num_stocks = conservative_stocks.size() + moderate_stocks.size() + aggressive_stocks.size();
            double weight = 1.0 / num_stocks;
            for (const auto& stock : conservative_stocks) {
                portfolio.emplace_back(stock.first, budget * weight);
            }
            for (const auto& stock : moderate_stocks) {
                portfolio.emplace_back(stock.first, budget * weight);
            }
            for (const auto& stock : aggressive_stocks) {
                portfolio.emplace_back(stock.first, budget * weight);
            }
            return portfolio;
        } else if (strategy == "random") {
            std::random_device rd;
            std::mt19937 gen(rd());
            std::gamma_distribution<> dist(1.0, 1.0);
            
            std::vector<double> weights;
            double sum = 0.0;
            for (size_t i = 0; i < 9; ++i) {
                double w = dist(gen);
                weights.push_back(w);
                sum += w;
            }
            
            for (double& w : weights) {
                w /= sum;
            }
            
            size_t idx = 0;
            for (const auto& stock : conservative_stocks) {
                portfolio.emplace_back(stock.first, budget * weights[idx++]);
            }
            for (const auto& stock : moderate_stocks) {
                portfolio.emplace_back(stock.first, budget * weights[idx++]);
            }
            for (const auto& stock : aggressive_stocks) {
                portfolio.emplace_back(stock.first, budget * weights[idx++]);
            }
            return portfolio;
        }
        
        // Allocate budget to each category
        double conservative_budget = budget * conservative_weight;
        double moderate_budget = budget * moderate_weight;
        double aggressive_budget = budget * aggressive_weight;
        
        // Distribute within each category
        for (const auto& stock : conservative_stocks) {
            portfolio.emplace_back(stock.first, conservative_budget / conservative_stocks.size());
        }
        
        for (const auto& stock : moderate_stocks) {
            portfolio.emplace_back(stock.first, moderate_budget / moderate_stocks.size());
        }
        
        for (const auto& stock : aggressive_stocks) {
            portfolio.emplace_back(stock.first, aggressive_budget / aggressive_stocks.size());
        }
        
        return portfolio;
    }

    void optimizePortfolio(const std::string& strategy) {
        // Initialize Prism client
        PrismClient client;
        
        // Get context from API
        auto [success, contextStr] = client.getContext();
        if (!success) {
            std::cerr << "Error getting context: " << contextStr << std::endl;
            return;
        }
        
        // Parse context
        parseContext(contextStr);
        
        // Calculate portfolio using selected strategy
        std::cout << "\nUsing strategy: " << strategy << std::endl;
        auto portfolio = calculateStockWeights(strategy);
        
        // Send portfolio to API
        auto [sendSuccess, response] = client.sendPortfolio(portfolio);
        if (!sendSuccess) {
            std::cerr << "Error submitting portfolio: " << response << std::endl;
            return;
        }
        
        try {
            double score = std::stod(response);
            std::cout << "Score: " << score << std::endl;
        } catch (const std::exception& e) {
            std::cerr << "Could not parse score from response: " << response << std::endl;
        }
        
        std::cout << "\nPortfolio allocation:" << std::endl;
        double total = 0.0;
        for (const auto& [stock, amount] : portfolio) {
            std::cout << stock << ": $" << std::fixed << std::setprecision(2) << amount << std::endl;
            total += amount;
        }
        std::cout << "Total: $" << std::fixed << std::setprecision(2) << total << std::endl;
    }
};

int main(int argc, char* argv[]) {
    // Define available strategies
    std::vector<std::string> available_strategies = {
        "balanced",     // Standard balanced approach based on risk tolerance
        "aggressive",   // Heavy focus on growth stocks
        "conservative", // Heavy focus on stable stocks
        "tech_heavy",   // Extreme focus on technology
        "equal_weight", // Equal distribution across all stocks
        "random"        // Random allocation
    };
    
    // Default strategy
    std::string strategy = "balanced";
    
    // Parse command line arguments
    if (argc > 1) {
        strategy = argv[1];
        if (std::find(available_strategies.begin(), available_strategies.end(), strategy) == available_strategies.end()) {
            std::cerr << "Invalid strategy. Available strategies are:" << std::endl;
            for (const auto& s : available_strategies) {
                std::cerr << "  " << s << std::endl;
            }
            return 1;
        }
    }
    
    // Run optimizer with selected strategy
    PortfolioOptimizer optimizer;
    optimizer.optimizePortfolio(strategy);
    
    return 0;
} 

File: portfolio_optimizer.py
import json
from typing import Dict, List, Tuple
import numpy as np
import argparse
from Prism import get_context, send_portfolio, get_my_current_information

class PortfolioOptimizer:
    def __init__(self):
        self.context = None
        self.risk_factors = {
            'age': 0.3,  # Further reduced age impact for maximum profit focus
            'employment_status': 0.1,  # Minimal employment impact
            'income': 0.6  # Increased income impact for maximum profit focus
        }
        self.sector_weights = {}  # Will be populated based on context
        self.budget = 0
        self.disliked_stocks = set()
        self.risk_tolerance = 0.5  # Default risk tolerance
        self.disliked_sectors = set()  # Sectors to avoid
        self.start_date = None
        self.end_date = None
        self.salary = 0

    def parse_context(self, context: str) -> None:
        """Parse the context string into structured data"""
        try:
            data = json.loads(context)
            self.context = data
            
            # Extract client preferences from message
            message = data.get('message', '')
            
            # Parse age
            age_str = message.split('is ')[1].split(' years')[0]
            self.age = int(age_str)
            
            # Parse budget
            budget_str = message.split('budget of $')[1].split('.')[0]
            self.budget = float(budget_str)
            
            # Parse dates
            start_date_str = message.split('started investing on ')[1].split(' and')[0]
            end_date_str = message.split('ended on ')[1].split('.')[0]
            self.start_date = start_date_str
            self.end_date = end_date_str
            
            # Parse salary
            salary_str = message.split('salary of $')[1].split('.')[0]
            self.salary = float(salary_str)
            
            # Parse disliked sectors
            if 'avoids' in message:
                disliked_sectors_str = message.split('avoids ')[1].split('.')[0]
                self.disliked_sectors = set(s.strip() for s in disliked_sectors_str.split(','))
            
            # Calculate risk tolerance based on client profile - adjusted for maximum profit focus
            age_factor = 1.0 - (min(1.0, self.age / 100))  # Invert age factor for more aggressive approach
            employment_factor = 0.9  # Higher for employed
            income_factor = min(1.0, self.salary / 80000.0)  # Lower salary threshold for more aggressive approach
            
            self.risk_tolerance = (
                age_factor * self.risk_factors['age'] +
                employment_factor * self.risk_factors['employment_status'] +
                income_factor * self.risk_factors['income']
            )
            
            # Ensure risk tolerance is between 0.5 and 0.95 for maximum profit focus
            self.risk_tolerance = max(0.5, min(0.95, self.risk_tolerance))
            
            print(f"Parsed context:")
            print(f"Age: {self.age}")
            print(f"Budget: ${self.budget}")
            print(f"Salary: ${self.salary}")
            print(f"Start Date: {self.start_date}")
            print(f"End Date: {self.end_date}")
            print(f"Disliked Sectors: {self.disliked_sectors}")
            print(f"Risk Tolerance: {self.risk_tolerance}")
            
        except (json.JSONDecodeError, IndexError, ValueError) as e:
            print(f"Error parsing context: {e}")
            print(f"Message content: {message}")
            raise

    def calculate_stock_weights(self, strategy: str = 'balanced') -> List[Tuple[str, float]]:
        """Calculate optimal stock weights based on constraints"""
        if not self.context:
            raise ValueError("Context not initialized")
            
        # Define sector classifications with maximum tech focus
        conservative_sectors = ['Utilities', 'Consumer Staples']
        aggressive_sectors = ['Technology', 'Communication Services', 'Consumer Discretionary']
        moderate_sectors = ['Financials', 'Industrials']
        
        # Updated stock selection with maximum profit focus
        conservative_stocks = [('JNJ', 1), ('PG', 1)]  # Reduced conservative stocks
        moderate_stocks = [('JPM', 1), ('MS', 1)]  # Reduced moderate stocks
        aggressive_stocks = [
            ('AAPL', 1), ('MSFT', 1), ('GOOGL', 1),
            ('NVDA', 1), ('TSLA', 1), ('META', 1),
            ('AMZN', 1), ('AMD', 1), ('INTC', 1),
            ('SNOW', 1), ('CRWD', 1), ('MDB', 1)  # Added more high-growth tech stocks
        ]
        
        portfolio = []
        
        # Add 1% buffer to prevent maxing out budget
        effective_budget = self.budget * 0.99
        
        if strategy == 'balanced':
            # Adjusted weights for maximum profit focus
            if self.risk_tolerance < 0.5:
                conservative_weight = 0.3
                moderate_weight = 0.2
                aggressive_weight = 0.5
            elif self.risk_tolerance > 0.7:
                conservative_weight = 0.05
                moderate_weight = 0.1
                aggressive_weight = 0.85
            else:
                conservative_weight = 0.1
                moderate_weight = 0.2
                aggressive_weight = 0.7
                
        elif strategy == 'aggressive':
            # Maximum aggressive allocation
            conservative_weight = 0.02
            moderate_weight = 0.08
            aggressive_weight = 0.9
            
        elif strategy == 'conservative':
            # Still maintaining significant growth focus
            conservative_weight = 0.5
            moderate_weight = 0.3
            aggressive_weight = 0.2
            
        elif strategy == 'tech_heavy':
            # Extreme focus on technology
            conservative_weight = 0.02
            moderate_weight = 0.03
            aggressive_weight = 0.95
            
        elif strategy == 'equal_weight':
            # Equal weight across all stocks
            num_stocks = len(conservative_stocks) + len(moderate_stocks) + len(aggressive_stocks)
            weight = 1.0 / num_stocks
            for stock, _ in conservative_stocks + moderate_stocks + aggressive_stocks:
                portfolio.append((stock, effective_budget * weight))
            return portfolio
            
        elif strategy == 'random':
            # Random allocation with maximum bias towards aggressive stocks
            weights = np.random.dirichlet(np.ones(16), size=1)[0]
            all_stocks = conservative_stocks + moderate_stocks + aggressive_stocks
            for (stock, _), weight in zip(all_stocks, weights):
                portfolio.append((stock, effective_budget * weight))
            return portfolio
            
        # Allocate budget to each category
        conservative_budget = effective_budget * conservative_weight
        moderate_budget = effective_budget * moderate_weight
        aggressive_budget = effective_budget * aggressive_weight
        
        # Distribute within each category
        for stock, _ in conservative_stocks:
            portfolio.append((stock, conservative_budget / len(conservative_stocks)))
            
        for stock, _ in moderate_stocks:
            portfolio.append((stock, moderate_budget / len(moderate_stocks)))
            
        for stock, _ in aggressive_stocks:
            portfolio.append((stock, aggressive_budget / len(aggressive_stocks)))
        
        return portfolio

    def optimize_portfolio(self, strategy: str) -> None:
        """Main optimization function"""
        # Get context from API
        success, context = get_context()
        if not success:
            print(f"Error getting context: {context}")
            return
            
        # Parse context
        self.parse_context(context)
        
        # Calculate portfolio using selected strategy
        print(f"\nUsing strategy: {strategy}")
        portfolio = self.calculate_stock_weights(strategy)
        
        # Send portfolio to API
        success, response = send_portfolio(portfolio)
        if success:
            try:
                score = float(response)
                print(f"Score: {score}")
            except ValueError:
                print(f"Could not parse score from response: {response}")
        else:
            print(f"Error submitting portfolio: {response}")
        
        print("\nPortfolio allocation:")
        total = 0.0
        for stock, amount in portfolio:
            print(f"{stock}: ${amount:.2f}")
            total += amount
        print(f"Total: ${total:.2f}")

def main():
    # Define available strategies
    available_strategies = [
        'balanced',    # Standard balanced approach based on risk tolerance
        'aggressive',  # Heavy focus on growth stocks
        'conservative', # Heavy focus on stable stocks
        'tech_heavy',  # Extreme focus on technology
        'equal_weight', # Equal distribution across all stocks
        'random'       # Random allocation
    ]
    
    # Set up argument parser
    parser = argparse.ArgumentParser(description='Portfolio Optimization')
    parser.add_argument('--strategy', type=str, choices=available_strategies,
                      default='balanced', help='Portfolio strategy to use')
    
    args = parser.parse_args()
    
    # Run optimizer with selected strategy
    optimizer = PortfolioOptimizer()

if __name__ == "__main__":
    main() 